src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,128 @@
+    mError                           = mpICDClientStorage->SetKey(mICDClientInfo, mNewKey.Span());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to store the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+    }
+}
+void ICDRefreshKeyInfo::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+}
+
+void ICDRefreshKeyInfo::OnDone(chip::app::CommandSender * apCommandSender)
+{
+    if (mError == CHIP_NO_ERROR)
+    {
+        mpCheckInDelegate->OnRegistrationUpdateComplete(mICDClientInfo, this);
+        return;
+    }
+    mpCheckInDelegate->OnRegistrationUpdateFailure(this);
+}
+
+CHIP_ERROR ICDRefreshKeyInfo::RegisterClientWithNewKey(Messaging::ExchangeManager & exchangeMgr,
+                                                       const SessionHandle & sessionHandle)
+{
+    using namespace Clusters::IcdManagement;
+
+    mRegisterCommandSender.SetExchangeManager(&exchangeMgr);
+    mRegisterCommandSender.SetCommandSenderDelegate(this);
+
+    auto commandPathParams = CommandPathParams(0, 0, Id, Commands::RegisterClient::Id, (CommandPathFlags::kEndpointIdValid));
+
+    ReturnErrorOnFailure(mRegisterCommandSender.PrepareCommand(commandPathParams));
+
+    chip::TLV::TLVWriter * writer = mRegisterCommandSender.GetCommandDataIBTLVWriter();
+
+    ReturnErrorOnFailure(
+        writer->Put(chip::TLV::ContextTag(Commands::RegisterClient::Fields::kCheckInNodeID), mICDClientInfo.peer_node.GetNodeId()));
+    ReturnErrorOnFailure(
+        writer->Put(chip::TLV::ContextTag(Commands::RegisterClient::Fields::kMonitoredSubject), mICDClientInfo.monitored_subject));
+    ReturnErrorOnFailure(writer->Put(chip::TLV::ContextTag(Commands::RegisterClient::Fields::kKey), mNewKey.Span()));
+
+    // TODO : We don't have plain data for the old key
+
+    ReturnErrorOnFailure(mRegisterCommandSender.FinishCommand());
+    ReturnErrorOnFailure(mRegisterCommandSender.SendCommandRequest(sessionHandle));
+
+    return CHIP_NO_ERROR;
+}
+
+void ICDRefreshKeyInfo::EstablishSessionToPeer()
+{
+    ChipLogProgress(ICD, "Trying to establish a CASE session for re-registering an ICD client");
+    auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
+    VerifyOrReturn(caseSessionManager != nullptr);
+    caseSessionManager->FindOrEstablishSession(mICDClientInfo.peer_node, &mOnConnectedCallback, &mOnConnectionFailureCallback);
+}
+
+void ICDRefreshKeyInfo::HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                              const SessionHandle & sessionHandle)
+{
+    ICDRefreshKeyInfo * const _this = static_cast<ICDRefreshKeyInfo *>(context);
+    VerifyOrDie(_this != nullptr);
+
+    if (CHIP_NO_ERROR != _this->RegisterClientWithNewKey(exchangeMgr, sessionHandle))
+    {
+        ChipLogError(ICD, "Failed to send register client command");
```

Comment:
Doesn't this leak?

Reply:
Re-registartion failure callback will be added here.